core/jobs/batch_jobs/story_node_jobs.py

```diff
@@ -0,0 +1,183 @@
+
+
+class PopulateStoryNodeJob(base_jobs.JobBase):
+    """Job that migrates story nodes, for serial chapter launch feature"""
+
+    DATASTORE_UPDATES_ALLOWED = True
+
+    def _update_story_node(
+            self, story_model: story_models.StoryModel
+    ) -> result.Result[
+        story_models.StoryModel,
+        Tuple[str, Exception]
+    ]:
+        """Populate the 5 new fields in each story node of the StoryModel
+        instance, namely: status, planned_publication_date_msecs,
+        last_modified_msecs, first_publication_date_msecs, unpublishing_reason.
+
+        Args:
+            story_model: StoryModel. The story whose nodes have to be
+                populated.
+
+        Returns:
+            Result(StoryModel, (str, Exception)). Result containing the
+            updated StoryModel instance to be put.
+        """
+
+        nodes = story_model.story_contents['nodes']
+
+        try:
+            with datastore_services.get_ndb_context():
+                topic_model = topic_models.TopicModel.get(
+                    story_model.corresponding_topic_id)
+                story_reference = next(
+                    story_ref for story_ref in (
+                        topic_model.canonical_story_references)
+                        if story_ref['story_id'] == story_model.id)
+                for node in nodes:
+                    node['unpublishing_reason'] = None
+                    node['status'] = 'Draft'
+                    if story_reference['story_is_published']:
+                        node['status'] = 'Published'
+
+                    current_topic_version = topic_model.version
+                    story_published_on = None
+                    for version in range(current_topic_version, 0, -1):
+                        snapshot_id = topic_model.get_snapshot_id(
+                            topic_model.id, version)
+                        topic_metadata = (
+                            topic_models.TopicSnapshotMetadataModel.get(
+                                snapshot_id))
+                        for cmd in topic_metadata.commit_cmds:
+                            if (cmd['cmd'] == 'publish_story' and
+                                cmd['story_id'] == story_model.id):
+                                story_published_on = (
+                                    utils.get_time_in_millisecs(
+                                    topic_metadata.created_on))
+                                break
+                        if story_published_on is not None:
+                            break
+
+                    current_story_version = story_model.version
+                    node_created_on = None
+                    for version in range(current_story_version, 0, -1):
+                        snapshot_id = story_model.get_snapshot_id(
+                                story_model.id, version)
+                        story_metadata = (
+                            story_models.StorySnapshotMetadataModel.get(
```

Comment:
Ditto*

Reply:
I have replied to that comment above. PTAL